pkg/virt-controller/watch/pool/pool.go

```diff
@@ -887,89 +955,138 @@ func (c *Controller) opportunisticUpdate(pool *poolv1.VirtualMachinePool, vmOutd
 	return nil
 }
 
-func (c *Controller) proactiveUpdate(pool *poolv1.VirtualMachinePool, vmUpdatedList []*virtv1.VirtualMachine) error {
-	var wg sync.WaitGroup
-	wg.Add(len(vmUpdatedList))
-	errChan := make(chan error, len(vmUpdatedList))
-	for i := 0; i < len(vmUpdatedList); i++ {
-		go func(idx int) {
-			defer wg.Done()
-			vm := vmUpdatedList[idx]
+func calculateMaxUnavailableInt(pool *poolv1.VirtualMachinePool) (int, error) {
+	maxUnavailable := intstr.FromString("100%")
+	if pool.Spec.MaxUnavailable != nil {
+		maxUnavailable = *pool.Spec.MaxUnavailable
+	}
 
-			vmiKey := controller.NamespacedKey(vm.Namespace, vm.Name)
-			obj, exists, _ := c.vmiStore.GetByKey(vmiKey)
-			if !exists {
-				// no VMI to update
-				return
-			}
-			vmi := obj.(*virtv1.VirtualMachineInstance)
-			if vmi.DeletionTimestamp != nil {
-				// ignore VMIs which are already deleting
-				return
-			}
+	totalReplicas := int32(1)
+	if pool.Spec.Replicas != nil {
+		totalReplicas = *pool.Spec.Replicas
+	}
 
-			updateType, err := c.isOutdatedVMI(vm, vmi)
-			if err != nil {
-				errChan <- err
-				return
+	var maxUnavailableInt int
+	if maxUnavailable.Type == intstr.String {
+		percentage, err := strconv.ParseInt(strings.TrimSuffix(maxUnavailable.StrVal, "%"), 10, 32)
+		if err != nil {
+			return 0, fmt.Errorf("invalid maxUnavailable percentage: %v", err)
+		}
+		maxUnavailableInt = int(float64(totalReplicas) * float64(percentage) / 100.0)
+	} else {
+		maxUnavailableInt = int(maxUnavailable.IntVal)
+	}
 
-			}
-			switch updateType {
-			case proactiveUpdateTypeRestart:
-				err := c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Delete(context.Background(), vmi.ObjectMeta.Name, v1.DeleteOptions{})
+	if maxUnavailableInt < 1 {
+		maxUnavailableInt = 1
+	}
+
+	return maxUnavailableInt, nil
+}
+
+func (c *Controller) proactiveUpdate(pool *poolv1.VirtualMachinePool, vmUpdatedList []*virtv1.VirtualMachine) error {
```

Comment:
There's way too much happening in this single function. Can you please break it down into smaller functions?